openpgp/s2k/s2k.go

```diff
@@ -14,56 +15,123 @@ import (
 	Hash crypto.Hash
-	// S2KCount is only used for symmetric encryption. It
+	// Argon2 parameters for S2K (String to Key).
+	// Only relevant if S2KMode is set to s2k.Argon2S2K.
+	// If nil, default parameters are used.
+	// For more details on the choice of parameters, see https://tools.ietf.org/html/rfc9106#section-4.
+	ArgonConf *ArgonConfig
+	// Only relevant if S2KMode is set to s2k.IterSaltedS2K.
+	// Iteration count for Iterated S2K (String to Key). It
 	// determines the strength of the passphrase stretching when
 	// the said passphrase is hashed to produce a key. S2KCount
-	// should be between 65536 and 65011712, inclusive. If Config
-	// is nil or S2KCount is 0, the value 16777216 used. Not all
+	// should be between 1024 and 65011712, inclusive. If Config
+	// is nil or S2KCount is 0, the value 65536 used. Not all
 	// values in the above range can be represented. S2KCount will
 	// be rounded up to the next representable value if it cannot
-	// be encoded exactly. See RFC 4880 Section 3.7.1.3.
+	// be encoded exactly. When set, it is strongly encrouraged to
+	// use a value that is at least 65536. See RFC 4880 Section
+	// 3.7.1.3.
 	S2KCount int
 }
 
+// ArgonConfig stores the Argon2 parameters
+// A nil *ArgonConfig is valid and results in all default
+type ArgonConfig struct {
+	NumberOfPasses      uint8
+	DegreeOfParallelism uint8
+	MemoryExponent      uint8
+}
+
 // Params contains all the parameters of the s2k packet
 type Params struct {
 	// mode is the mode of s2k function.
 	// It can be 0 (simple), 1(salted), 3(iterated)
 	// 2(reserved) 100-110(private/experimental).
-	mode uint8
+	mode S2KType
 	// hashId is the ID of the hash function used in any of the modes
 	hashId byte
-	// salt is a byte array to use as a salt in hashing process
+	// salt is a byte array to use as a salt in hashing process or argon2
 	salt []byte
 	// countByte is used to determine how many rounds of hashing are to
 	// be performed in s2k mode 3. See RFC 4880 Section 3.7.1.3.
 	countByte byte
+	// passes is a parameter in Argon2 to determine the number of iterations
+	// See RFC 4880 Section 3.7.1.4.
+	passes byte
+	// parallelism is a parameter in Argon2 to determine the degree of paralellism
+	// See RFC 4880 Section 3.7.1.4.
+	parallelism byte
+	// memoryExp is a parameter in Argon2 to determine the memory usage
+	// i.e., 2 pow memoryExp kibibytes
+	// See RFC 4880 Section 3.7.1.4.
+	memoryExp byte
 }
 
-func (c *Config) hash() crypto.Hash {
+func (c *S2KConfig) Mode() S2KType {
+	if c == nil {
+		return IterSaltedS2K
+	}
+	return c.S2KMode
+}
+
+func (c *S2KConfig) hash() crypto.Hash {
 	if c == nil || uint(c.Hash) == 0 {
 		return crypto.SHA256
 	}
 
 	return c.Hash
 }
 
+func (c *S2KConfig) ArgonConfig() *ArgonConfig {
+	if c == nil || c.ArgonConf == nil {
+		return DefaultArgonConfig()
+	}
+	return c.ArgonConf
+}
+
+func (c *S2KConfig) Count() int {
+	if c == nil || c.S2KCount == 0 {
+		return 65536
+	}
+	return c.S2KCount
+}
```

Comment:
Do we need this function? It seems unused, and it might be better to have only one way to get the count (via `EncodedCount()` below) - as currently they don't match.